File: Chapter03/Component.cpp
//
// Created by Michael Lehmann on 10.01.21.
//

#include "Component.h"
#include "Actor.h"

Component::Component(Actor* owner, int updateOrder)
        :mOwner(owner)
        ,mUpdateOrder(updateOrder)
{
    // Add to actor's vector of components
    mOwner->AddComponent(this);
}

Component::~Component()
{
    mOwner->RemoveComponent(this);
}

void Component::Update(float deltaTime)
{
}

File: Chapter06_Veldrid/Component.cs
using System;
using Veldrid;

namespace Chapter06
{
    public abstract class Component : IDisposable
    {
        protected Component(Actor owner, int updateOrder = 100)
        {
            Owner = owner;
            UpdateOrder = updateOrder;

            // Add to actor's vector of components
            Owner.AddComponent(this);
        }

        public Actor Owner { get; }

        public int UpdateOrder { get; }

        public virtual void Update(float deltaTime)
        {
        }

        public virtual void ProcessInput(InputState input)
        {
        }

        public virtual void OnUpdateWorldTransform()
        {
        }

        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }
        
        protected virtual void Dispose(bool disposing)
        {
            if (disposing)
            {
                Owner.RemoveComponent(this);
            }
        }
    }
}